src/nvim/fileio.c

```diff
@@ -5239,7 +5239,14 @@ int delete_recursive(char_u *name)
   int result = 0;
 
   if (os_isrealdir(name)) {
-    snprintf((char *)NameBuff, MAXPATHL, "%s/*", name);  // NOLINT
+    // Escape some regexp chars in file names as these probably will be
+    // interpreted as ill-formed regular expressions in gen_expand_wildcards.
+    char_u *esc_name = do_string_sub(name, (char_u *)"[",
+                                     (char_u *)"[[]", (char_u *)"g");
+    char_u *esc_name2 = vim_strsave_escaped(esc_name, (char_u *)"{}?");
```

Comment:
For Windows (`#ifdef WIN32`) try this:

    char_u *esc_name2 = vim_strsave_escaped_ext(esc_name, "{}", '^', false);
 
if that doesn't work, maybe this...

    char_u *esc_name2 = do_string_sub(esc_name, (char_u *)"{", (char_u *)"\\^{", (char_u *)"g");
    esc_name2 = do_string_sub(esc_name, (char_u *)"}", (char_u *)"\\^}", (char_u *)"g");

(`^` is the win32 shell escape char; though I don't see anything in `gen_expand_wildcards` that handles it, IIRC `gen_expand_wildcards` may defer to the OS shell for wildcard expansion in some cases... didn't look closely)